packages/idempotency/src/IdempotencyHandler.ts

```diff
@@ -1,41 +1,84 @@
-} from './types';
 import {
+  IdempotencyAlreadyInProgressError,
   IdempotencyInconsistentStateError,
   IdempotencyItemAlreadyExistsError,
-  IdempotencyAlreadyInProgressError,
-  IdempotencyPersistenceLayerError
+  IdempotencyPersistenceLayerError,
 } from './Exceptions';
-import { IdempotencyRecord } from './persistence/IdempotencyRecord';
+import { IdempotencyRecord } from './persistence';
 
 export class IdempotencyHandler<U> {
   public constructor(
     private functionToMakeIdempotent: AnyFunctionWithRecord<U>,
-    private functionPayloadToBeHashed: Record<string, unknown>, 
+    private functionPayloadToBeHashed: Record<string, unknown>,
     private idempotencyOptions: IdempotencyOptions,
-    private fullFunctionPayload: Record<string, unknown>
-  ) {}
+    private fullFunctionPayload: Record<string, unknown>,
+  ) {
+  }
 
-  public determineResultFromIdempotencyRecord(idempotencyRecord: IdempotencyRecord): Promise<U> | U { 
+  public determineResultFromIdempotencyRecord(
+    idempotencyRecord: IdempotencyRecord
+  ): Promise<U> | U {
     if (idempotencyRecord.getStatus() === IdempotencyRecordStatus.EXPIRED) {
-      throw new IdempotencyInconsistentStateError('Item has expired during processing and may not longer be valid.');
-    } else if (idempotencyRecord.getStatus() === IdempotencyRecordStatus.INPROGRESS){
-      throw new IdempotencyAlreadyInProgressError(`There is already an execution in progress with idempotency key: ${idempotencyRecord.idempotencyKey}`);
+      throw new IdempotencyInconsistentStateError(
+        'Item has expired during processing and may not longer be valid.'
+      );
+    } else if (
+      idempotencyRecord.getStatus() === IdempotencyRecordStatus.INPROGRESS
+    ) {
+      if (
+        idempotencyRecord.inProgressExpiryTimestamp &&
+        idempotencyRecord.inProgressExpiryTimestamp <
+        new Date().getUTCMilliseconds()
+      ) {
+        throw new IdempotencyInconsistentStateError(
+          'Item is in progress but the in progress expiry timestamp has expired.'
+        );
+      } else {
+        throw new IdempotencyAlreadyInProgressError(
+          `There is already an execution in progress with idempotency key: ${idempotencyRecord.idempotencyKey}`
+        );
+      }
     } else {
       // Currently recalling the method as this fulfills FR1. FR3 will address using the previously stored value https://github.com/awslabs/aws-lambda-powertools-typescript/issues/447
-      return this.functionToMakeIdempotent(this.fullFunctionPayload); 
+      return this.functionToMakeIdempotent(this.fullFunctionPayload);
+    }
+  }
+
+  /**
+   * Main entry point for the handler
+   * IdempotencyInconsistentStateError can happen under rare but expected cases
+   * when persistent state changes in the small time between put & get requests.
+   * In most cases we can retry successfully on this exception.
+   */
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
```

Comment:
Before we merge, can we fix this return type and remove the comment?

Based on what I see, we either need to change the return type to `Promise<U | undefined>` or make sure that the function returns something in the `catch` block when the `if` statement is skipped.

Reply:
Hmm this is  weird construct we have there. Adding `undefined` to the Promise would leak it further to `makeIdempotentFunction` and also higher to the user api, which I'd like to avoid. TS complains about missing return statement outside of the for-loop. But adding a statement makes it impossible to test, because this return statement is unreachable. I have added a throw with coverage ignore.